synapse/storage/databases/main/devices.py

```diff
@@ -1045,6 +1050,62 @@ def _txn(txn):
 
         return rows
 
+    async def _remove_deleted_devices_from_device_inbox(
+        self, progress: JsonDict, batch_size: int
+    ) -> int:
+        """A background update that deletes all device_inboxes for deleted devices.
+
+        This should only need to be run once (when users upgrade to v1.45.0)
+
+        Args:
+            progress: JsonDict used to store progress of this background update
+            batch_size: the maximum number of rows to retrieve in a single select query
+
+        Returns:
+            The number of deleted rows
+        """
+
+        def _remove_deleted_devices_from_device_inbox_txn(
+            txn: LoggingTransaction,
+        ) -> int:
+
+            sql = """
+                SELECT user_id, device_id, stream_id
+                    FROM device_inbox
+                    WHERE device_id
+                        NOT IN (SELECT device_id FROM devices)
+                    LIMIT ?;
+            """
+
+            txn.execute(sql, (batch_size,))
+            rows = txn.fetchall()
+
+            for row in rows:
+                self.db_pool.simple_delete_txn(
+                    txn,
+                    "device_inbox",
+                    {"user_id": row[0], "device_id": row[1], "stream_id": row[2]},
+                )
+
+            if rows:
+                self.db_pool.updates._background_update_progress_txn(
+                    txn, "remove_deleted_devices_from_device_inbox", row
+                )
```

Comment:
We don't seem to use the progress we've made at all here, it gets passed back into `_remove_deleted_devices_from_device_inbox` but we seem to always run the same query. We might want to do something like `WHERE device_id > last_seen_device_id`.